### ISSUE TITLE
[no-explicit-any] Option to allow `any` in the definition of generic type arguments

### ISSUE THREAD

**Issue by tjespe** [typescript-eslint/typescript-eslint#3038]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-explicit-any": "on"
  }
}
```

```TS
const myGenericFunction = <T extends Record<string, any>>(argument: T) =>
  Object.fromEntries(
    Object.entries(argument).map(([key, value]) => [key + " modified", value])
  );
```

**Expected Result**
I would expect that using any in this case would be allowed, because as long as the argument is typed in the code that calls the function it is completely type safe as far as I know.

It would be great to at least have an option to allow using any types in the definition of generic type arguments, because it is pretty hard to write good generic functions with some type constraints with the current behavior of this rule.

**Actual Result**

Got error: `Unexpected any. Specify a different type.`

**Additional Info**

No exception.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.12.0` |
| `@typescript-eslint/parser`        | `4.12.0` |
| `TypeScript`                       | `4.1.3` |
| `ESLint`                           | `7.17.0` |
| `node`                             | `14.12.0` |

**Comment by bradzacher** [comment]
> I would expect that using any in this case would be allowed, because as long as the argument is typed in the code that calls the function it is completely type safe as far as I know.

This is incorrect.

The types of generic constraints are carried through into the function, and TS will use those constraints as the base type of the variable.

This means that in your example `value` is typed as `any`.
Which is entirely unsafe!!!

Here's an example which typechecks fine due to the `any`, but is *unsafe* and crashes at runtime:

```ts
const myGenericFunction = <T extends Record<string, any>>(argument: T) => {
  const lengths = [];
  for (const key of Object.keys(argument)) {
    const value = argument[key]; // typeof value === any
    lengths.push(value.length.toPrecision(2));
  }
  return lengths;
};

myGenericFunction({a: 1, b: 'str', c: false});
```
[repl](https://www.typescriptlang.org/play?target=99&ssl=1&ssc=1&pln=10&pc=47#code/MYewdgzgLgBAtgTwOIFMwoE4EtgDECuYwUW4MAvDADwAqMKAHlGgCYQwBKKoGLV02MAHMANDACGYBAD5pACnEYh+OGigAuGDQCUFaTADeAKBgxQkWABs0QqAAt2lANoBdANwmYAMxAYYc82gYAGsUBBgQLxgAeQAjACtuKAA6UIQIBSUVNW1dY1NTQNgAN3FLfBQKCSzVMCgnNPcYAHpmmCgEAAcUSJhS8sryIYkpT1NrYXsIZM78CDs5forkids7ZKgQAAUMbiwIUjA5ACZcj1MAX09dqHwMMBhVqY8LjyNEVHRsPEJiQ7kDOJNABGMSxTQAcgEELEwE0XjKEBQF20biAA)

For most cases can substitute the `any` for an `unknown`. For almost all other cases you can substitute `any` for `never`.
There are some *rare* cases that you have to use `any` - but they are exactly that - RARE.
